### PR TITLE
refactor: remove duplicated code

### PR DESCRIPTION
**Pull request by localvar**
Major changes:
* move `Proxy` and `GRPCProxy` into package `proxies`
* move service discovery to `proxies`
* move load balance to `proxies`
* move request match to `proxies`

### CODE REVIEW

**Comment by xxx7xxxx** [comment]
It's hard to review such a huge PR,  could you please point out the change of logic parts besides moves.

**Review by grootpiano — COMMENTED** [review]
Awesome PR! Learned a lot. So as return, I pointed out one comment typo, hhh 😄

**Comment by localvar** [comment]
> It's hard to review such a huge PR, could you please point out the change of logic parts besides moves.

there are no logic changes, but to extract the duplicated code, I've made a big refactor to the implementation of load balancers, by splitting it into LoadBalancer, SessionSticker, and HealthyChecker, this is the major changes.

File: pkg/object/grpcserver/spec.go [hunk 1]
```diff
@@ -94,7 +95,7 @@ func (h *Header) initHeaderRoute() {
 }
 
 // Validate validates Path.
```
Comment (suchen-sci) [inline pkg/object/grpcserver/spec.go:97]:
```suggestion
// Validate validates Method.
```

**Review by suchen-sci — COMMENTED** [review]
By the way, do you consider to update the doc about grpc `doc/cookbook/grpc-server.md`?

**Review by suchen-sci — APPROVED** [review]

**Review by grootpiano — APPROVED** [review]

File: pkg/object/grpcserver/spec.go [hunk 2]
```diff
@@ -63,15 +64,15 @@ type (
...
-		Paths      []*Path        `json:"paths" jsonschema:"omitempty"`
+		Methods    []*Method      `json:"methods" jsonschema:"omitempty"`
 	}
 
-	// Path is second level entry of router.
-	Path struct {
+	// Method is second level entry of router.
+	Method struct {
 		IPFilter       *ipfilter.Spec `json:"ipFilter,omitempty" jsonschema:"omitempty"`
-		Path           string         `json:"path,omitempty" jsonschema:"omitempty,pattern=^/"`
```
Comment (xxx7xxxx) [inline pkg/object/grpcserver/spec.go:72]: Why delete `path`, am I missing something?

Comment (localvar) [reply pkg/object/grpcserver/spec.go:72]: this is the grpc server, not deleting `path` but renaming it to `method`.

**Review by xxx7xxxx — APPROVED** [review]

**Comment by xxx7xxxx** [comment]
Need a conflicting fix.

**Pull request merged**